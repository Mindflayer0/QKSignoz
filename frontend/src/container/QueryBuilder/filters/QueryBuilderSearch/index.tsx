import { Select, Tag } from 'antd';
import { useAutoComplete } from 'hooks/queryBuilder/useAutoComplete';
import React, { useCallback } from 'react';
import { IBuilderQuery } from 'types/api/queryBuilder/queryBuilderData';

import { filterSelectStyle } from './config';

function QueryBuilderSearch({ query }: QueryBuilderSearchProps): JSX.Element {
	const {
		handleClearTag,
		handleKeyDown,
		handleSearch,
		handleSelect,
		tags,
		options,
		searchValue,
		isFilter,
	} = useAutoComplete(query);

	const onTagRender = ({ value }: { value: string }): React.ReactElement => (
		<Tag closable>{value}</Tag>
	);

	const getOptionClasses = useCallback(
		(selected: boolean | undefined) =>
			selected ? 'ant-select-item-option-selected select-item-option-state' : '',
		[],
	);

	return (
		<Select
			showSearch
			tagRender={onTagRender}
			filterOption={isFilter}
			autoClearSearchValue={false}
			mode="multiple"
			placeholder="Search Filter"
			value={tags}
			disabled={!query.aggregateAttribute}
			style={filterSelectStyle}
			onDeselect={handleClearTag}
			onSelect={handleSelect}
			onInputKeyDown={handleKeyDown}
			onSearch={handleSearch}
			searchValue={searchValue}
		>
			{options?.map((option) => (
				<Select.Option
					key={option.value}
					value={option.value}
					className={getOptionClasses(option.selected)}
				>
					{option.value}
				</Select.Option>
			))}
		</Select>
	);
}

interface QueryBuilderSearchProps {
	query: IBuilderQuery;
}

export default QueryBuilderSearch;
