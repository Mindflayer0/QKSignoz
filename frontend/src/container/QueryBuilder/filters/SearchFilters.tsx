import { Select } from 'antd';
import { IOption } from 'container/MetricsApplication/ResourceAttributesFilter/types';
import React, { useCallback, useEffect, useState } from 'react';

function SearchFilters(): JSX.Element {
	const [currentState, setCurrentState] = useState<string>('tagKey');
	const [optionsData, setOptionsData] = useState<IOption[]>([]);

	useEffect(() => {
		if (currentState === 'tagKey') {
			setOptionsData([
				{ value: 'jack', label: 'Jack' },
				{ value: 'lucy', label: 'Lucy' },
				{ value: 'Yiminghe', label: 'yiminghe' },
				{ value: 'disabled', label: 'Disabled' },
			]);
		}
		if (currentState === 'tagOperator') {
			setOptionsData([
				{ value: '=', label: '=' },
				{ value: '!=', label: '!=' },
				{ value: '>', label: '>' },
				{ value: '<', label: '<' },
			]);
		}
		if (currentState === 'tagValue') {
			setOptionsData([
				{ value: '12', label: '12' },
				{ value: '34', label: '34' },
				{ value: '10', label: '10' },
				{ value: '14', label: '14' },
			]);
		}
	}, [currentState]);

	const handleChange = useCallback(
		(value: string[]): void => {
			console.log(`selected ${value}`);
			if (currentState === 'tagKey') {
				setCurrentState('tagOperator');
			}
			if (currentState === 'tagOperator') {
				setCurrentState('tagValue');
			}
			if (currentState === 'tagValues') {
				setCurrentState('');
			}
		},
		[currentState],
	);

	return (
		<Select
			allowClear
			placeholder="Type your Query"
			onChange={handleChange}
			style={{ width: '100%' }}
			mode="tags"
			showArrow={false}
			optionLabelProp="label"
			options={optionsData}
		/>
	);
}

export default SearchFilters;
