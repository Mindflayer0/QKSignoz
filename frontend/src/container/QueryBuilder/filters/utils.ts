import { HAVING_FILTER_REGEXP } from 'constants/regExp';
import { HavingForm } from 'types/api/queryBuilder/queryBuilderData';

export const handleKeyDownLimitFilter: React.KeyboardEventHandler<HTMLInputElement> = (
	event,
): void => {
	const keyCode = event.keyCode || event.which;
	const isBackspace = keyCode === 8;
	const isNumeric =
		(keyCode >= 48 && keyCode <= 57) || (keyCode >= 96 && keyCode <= 105);

	if (!isNumeric && !isBackspace) {
		event.preventDefault();
	}
};

export const getHavingObject = (currentSearch: string): HavingForm => {
	const textArr = currentSearch.split(' ');
	const [columnName = '', op = '', ...value] = textArr;

	return { columnName, op, value };
};

export const isValidHavingValue = (search: string): boolean => {
	const values = getHavingObject(search).value.join(' ');

	if (values) {
		return HAVING_FILTER_REGEXP.test(values);
	}

	return true;
};
