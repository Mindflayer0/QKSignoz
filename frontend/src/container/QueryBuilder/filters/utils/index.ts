import {
	AttributeKeyOptions,
	getAttributesKeys,
	getAttributesValues,
	TagKeyValueProps,
} from 'api/queryBuilder/getAttributesKeysValues';
import { IOption } from 'container/FormAlertRules/labels/types';

export const GetTagKeys = async (
	body: TagKeyValueProps,
): Promise<IOption[]> => {
	const { payload } = await getAttributesKeys(body);
	if (!payload || !payload?.data) {
		return [];
	}
	return payload.data.attributeKeys.map((tagKey: AttributeKeyOptions) => ({
		label: tagKey.key,
		value: tagKey.key,
	}));
};

export const GetTagValues = async (
	attributeKey: string,
	dataSource: string,
): Promise<IOption[]> => {
	const { payload } = await getAttributesValues({
		attributeKey,
		dataSource,
	});

	if (!payload || !payload?.data) {
		return [];
	}
	return payload.data.attributeValues.map((tagKey: AttributeKeyOptions) => ({
		label: tagKey.key,
		value: tagKey.key,
	}));
};
