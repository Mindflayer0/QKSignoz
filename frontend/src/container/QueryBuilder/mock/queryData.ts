import { initialQueryBuilderFormValuesMap } from 'constants/queryBuilder';
import { IBuilderQuery } from 'types/api/queryBuilder/queryBuilderData';
import { DataSource } from 'types/common/queryBuilder';

export const queryMockData: IBuilderQuery = {
	dataSource: DataSource.METRICS,
	queryName: 'A',
	aggregateOperator: 'noop',
	aggregateAttribute: {
		key: 'signoz_latency_count',
		dataType: 'float64',
		type: 'tag',
		isColumn: true,
	},
	filters: {
		items: [
			{
				id: '8fac746b',
				key: {
					key: 'resource_signoz_collector_id',
					dataType: 'int64',
					isColumn: false,
					type: 'tag',
				},
				op: '=',
				value: ['1a5d3cc2-4b3e-4c7c-ad07-c4cdd739d1b9'],
			},
			{
				id: '8fadtr46b',
				key: {
					key: 'service_name',
					dataType: 'string',
					isColumn: false,
					type: 'tag',
				},
				op: 'IN',
				value: ['frontend', 'operation', 'ser'],
			},
		],
		op: 'AND',
	},
	expression: '',
	disabled: false,
	having: [],
	stepInterval: 30,
	limit: 10,
	orderBy: [],
	groupBy: [],
	legend: '',
	reduceTo: 'sum',
};

export const formulaMockData = {
	queryName: 'C',
	label: 'Formula',
	expression: 'sum(A, B)',
	legend: 'Total',
	disabled: false,
};

export const valueWithAttributeAndOperator: IBuilderQuery = {
	...initialQueryBuilderFormValuesMap.logs,
	dataSource: DataSource.LOGS,
	aggregateOperator: 'SUM',
	aggregateAttribute: {
		isColumn: false,
		key: 'bytes',
		type: 'tag',
		dataType: 'float64',
	},
};

export const querySearchResult = ['frontend', 'operation', 'ser'];
