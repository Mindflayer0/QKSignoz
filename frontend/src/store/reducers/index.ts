import { combineReducers } from 'redux';

import appReducer from './app';
import dashboardReducer from './dashboard';
import globalTimeReducer from './global';
import { LogsReducer } from './logs';
import metricsReducers from './metric';
import { PipelineReducer } from './pipeline';
import { ServiceMapReducer } from './serviceMap';
import traceReducer from './trace';
import { usageDataReducer } from './usage';

const reducers = combineReducers({
	traces: traceReducer,
	usageDate: usageDataReducer,
	globalTime: globalTimeReducer,
	serviceMap: ServiceMapReducer,
	dashboards: dashboardReducer,
	app: appReducer,
	metrics: metricsReducers,
	logs: LogsReducer,
	pipeline: PipelineReducer,
});

export type AppState = ReturnType<typeof reducers>;

export default reducers;
