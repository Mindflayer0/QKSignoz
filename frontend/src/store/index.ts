import { applyMiddleware, compose, createStore } from 'redux';
import type { ThunkMiddleware } from 'redux-thunk';
import thunk from 'redux-thunk';

import reducers, { AppState } from './reducers';

const composeEnhancers =
	(window && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const store = createStore(
	reducers,
	// @TODO Add Type for AppActions also
	composeEnhancers(applyMiddleware(thunk as ThunkMiddleware<AppState>)),
);

export default store;
