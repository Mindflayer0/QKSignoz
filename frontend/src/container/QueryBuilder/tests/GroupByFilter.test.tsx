import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import i18n from 'ReactI18';
import store from 'store';

import { GroupByFilter } from '../filters';
import { queryMockData } from '../mock/queryData';

describe('QueryBuilder: groupBy Filter', () => {
	const queryClient = new QueryClient({
		defaultOptions: {
			queries: {
				refetchOnWindowFocus: false,
			},
		},
	});

	it('it should render GroupByFilter', () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Provider store={store}>
					<I18nextProvider i18n={i18n}>
						<QueryClientProvider client={queryClient}>
							<GroupByFilter
								onChange={jest.fn()}
								query={queryMockData}
								disabled={false}
							/>
						</QueryClientProvider>
					</I18nextProvider>
				</Provider>
			</MemoryRouter>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
