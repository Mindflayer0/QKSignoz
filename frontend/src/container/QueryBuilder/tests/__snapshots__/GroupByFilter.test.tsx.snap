// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder: groupBy Filter it should render GroupByFilter 1`] = `
<DocumentFragment>
  <div
    class="ant-select css-dev-only-do-not-override-1i536d8 ant-select-multiple ant-select-show-search"
    style="width: 100%; min-width: 7.7rem;"
  >
    <div
      class="ant-select-selector"
    >
      <div
        class="ant-select-selection-overflow"
      >
        <div
          class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
          style="opacity: 1;"
        >
          <div
            class="ant-select-selection-search"
            style="width: 0px;"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              role="combobox"
              type="search"
              value=""
            />
            <span
              aria-hidden="true"
              class="ant-select-selection-search-mirror"
            >
              Â 
            </span>
          </div>
        </div>
      </div>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
  </div>
</DocumentFragment>
`;
